### PR TITLE
Consistently use /_not-found for not found page in App Router

### PR DESCRIPTION
**Pull request by timneutkens**
## What?

#62528 caused test/e2e/app-dir/not-found/conflict-route to fail compilation in Turbopack, this compiler error was previously already reported by Turbopack but Next.js didn't show it, which #62528 resolved.

This PR changes the handling for the not-found handling to be consistent between development and build, which ensures that the "special" page no longer conflicts with app/not-found/page.js.

Closes NEXT-2617

Note: this is a reworked iteration of https://github.com/vercel/next.js/pull/62585 which wasn't sufficient.

### CODE REVIEW

**Comment by timneutkens** [comment]
* **#62679** <a href="https://app.graphite.dev/github/pr/vercel/next.js/62679?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `canary`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @timneutkens and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 2864324c867cd963d0e03f3cf72fa9308fd409be

`pnpm test-dev test/e2e/app-dir/not-found-default/index.test.ts` (PPR)
- app dir - not found with default 404 page > should render default 404 with root layout for non-existent page

<details>
<summary>Expand output</summary>

  ● app dir - not found with default 404 page › should render default 404 with root layout for non-existent page

    expect(received).toContain(expected) // indexOf

    Expected substring: "/not-found"
    Received string:    "yarn run v1.22.19
    $ /tmp/next-install-2928c798a7d39724cbe89f6596ce98dc73f2cb0633833bcb3550b0946f78c384/node_modules/.bin/next
     ⚠ `experimental.ppr` has been defaulted to `true` because `__NEXT_EXPERIMENTAL_PPR` was set to `true` during testing.
     ⚠ `experimental.ppr` has been defaulted to `true` because `__NEXT_EXPERIMENTAL_PPR` was set to `true` during testing.
       ▲ Next.js 14.1.1-canary.80
       - Local:        http://localhost:46527·
     ⚠ `experimental.ppr` has been defaulted to `true` because `__NEXT_EXPERIMENTAL_PPR` was set to `true` during testing.
     ⚠ `experimental.ppr` has been defaulted to `true` because `__NEXT_EXPERIMENTAL_PPR` was set to `true` during testing.
     ✓ Ready in 1676ms
     ○ Compiling / ...
     ✓ Compiled / in 3s (480 modules)
     ✓ Compiled /_not-found in 327ms (483 modules)
    "

      35 |       if (isNextDev) {
      36 |         const cliOutput = next.cliOutput
    > 37 |         expect(cliOutput).toContain('/not-found')
         |                           ^
      38 |         expect(cliOutput).not.toContain('/_error')
      39 |       }
      40 |     })

      at Object.toContain (e2e/app-dir/not-found-default/index.test.ts:37:27)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong></summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 02-29-Consistently_use_/_not-found_for_not_found_page_in_App_Router | Change |
| - | - | - | - |
| buildDuration | 13.8s | 14s | ⚠️ +167ms |
| buildDurationCached | 7.5s | 6.3s | N/A |
| nodeModulesSize | 197 MB | 197 MB | N/A |
| nextStartRea..uration (ms) | 406ms | 407ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 02-29-Consistently_use_/_not-found_for_not_found_page_in_App_Router | Change |
| - | - | - | - |
| 2453-HASH.js gzip | 30.5 kB | 30.4 kB | N/A |
| 3304.HASH.js gzip | 181 B | 181 B | ✓ |
| 3f784ff6-HASH.js gzip | 53.7 kB | 53.7 kB | N/A |
| 8299-HASH.js gzip | 5.04 kB | 5.04 kB | N/A |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 242 B | 242 B | ✓ |
| main-HASH.js gzip | 32.1 kB | 32.2 kB | N/A |
| webpack-HASH.js gzip | 1.7 kB | 1.7 kB | N/A |
| Overall change | 45.6 kB | 45.6 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 02-29-Consistently_use_/_not-found_for_not_found_page_in_App_Router | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 02-29-Consistently_use_/_not-found_for_not_found_page_in_App_Router | Change |
| - | - | - | - |
| _app-HASH.js gzip | 196 B | 197 B | N/A |
| _error-HASH.js gzip | 184 B | 184 B | ✓ |
| amp-HASH.js gzip | 505 B | 505 B | ✓ |
| css-HASH.js gzip | 324 B | 325 B | N/A |
| dynamic-HASH.js gzip | 2.5 kB | 2.5 kB | N/A |
| edge-ssr-HASH.js gzip | 258 B | 258 B | ✓ |
| head-HASH.js gzip | 352 B | 352 B | ✓ |
| hooks-HASH.js gzip | 370 B | 371 B | N/A |
| image-HASH.js gzip | 4.2 kB | 4.2 kB | ✓ |
| index-HASH.js gzip | 259 B | 259 B | ✓ |
| link-HASH.js gzip | 2.67 kB | 2.67 kB | N/A |
| routerDirect..HASH.js gzip | 314 B | 312 B | N/A |
| script-HASH.js gzip | 386 B | 386 B | ✓ |
| withRouter-HASH.js gzip | 309 B | 309 B | ✓ |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 6.56 kB | 6.56 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 02-29-Consistently_use_/_not-found_for_not_found_page_in_App_Router | Change |
| - | - | - | - |
| _buildManifest.js gzip | 483 B | 485 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 02-29-Consistently_use_/_not-found_for_not_found_page_in_App_Router | Change |
| - | - | - | - |
| index.html gzip | 529 B | 527 B | N/A |
| link.html gzip | 542 B | 540 B | N/A |
| withRouter.html gzip | 524 B | 521 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 02-29-Consistently_use_/_not-found_for_not_found_page_in_App_Router | Change |
| - | - | - | - |
| edge-ssr.js gzip | 95.1 kB | 95.1 kB | N/A |
| page.js gzip | 3.06 kB | 3.06 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 02-29-Consistently_use_/_not-found_for_not_found_page_in_App_Router | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 624 B | 624 B | ✓ |
| middleware-r..fest.js gzip | 151 B | 151 B | ✓ |
| middleware.js gzip | 25.4 kB | 25.5 kB | N/A |
| edge-runtime..pack.js gzip | 839 B | 839 B | ✓ |
| Overall change | 1.61 kB | 1.61 kB | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 02-29-Consistently_use_/_not-found_for_not_found_page_in_App_Router | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 171 kB | 171 kB | N/A |
| app-page-exp..prod.js gzip | 96.7 kB | 96.7 kB | N/A |
| app-page-tur..prod.js gzip | 98.5 kB | 98.5 kB | N/A |
| app-page-tur..prod.js gzip | 92.9 kB | 92.9 kB | N/A |
| app-page.run...dev.js gzip | 150 kB | 150 kB | N/A |
| app-page.run..prod.js gzip | 91.4 kB | 91.4 kB | N/A |
| app-route-ex...dev.js gzip | 21.3 kB | 21.3 kB | ✓ |
| app-route-ex..prod.js gzip | 15 kB | 15 kB | ✓ |
| app-route-tu..prod.js gzip | 15 kB | 15 kB | ✓ |
| app-route-tu..prod.js gzip | 14.8 kB | 14.8 kB | ✓ |
| app-route.ru...dev.js gzip | 20.9 kB | 20.9 kB | ✓ |
| app-route.ru..prod.js gzip | 14.8 kB | 14.8 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.51 kB | 9.51 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.79 kB | 9.79 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.51 kB | 9.51 kB | ✓ |
| pages-turbo...prod.js gzip | 22.3 kB | 22.3 kB | ✓ |
| pages.runtim...dev.js gzip | 23 kB | 23 kB | ✓ |
| pages.runtim..prod.js gzip | 22.3 kB | 22.3 kB | ✓ |
| server.runti..prod.js gzip | 50.6 kB | 50.7 kB | N/A |
| Overall change | 198 kB | 198 kB | ✓ |

</details>

<details>
<summary><strong>build cache</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 02-29-Consistently_use_/_not-found_for_not_found_page_in_App_Router | Change |
| - | - | - | - |
| 0.pack gzip | 1.56 MB | 1.56 MB | N/A |
| index.pack gzip | 105 kB | 105 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>middleware.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>edge-ssr.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>2453-HASH.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>main-HASH.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page-exp..ntime.dev.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page-exp..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page-tur..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page-tur..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page.runtime.dev.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page.runtime.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>server.runtime.prod.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: 2864324c867cd963d0e03f3cf72fa9308fd409be

File: packages/next-swc/crates/next-core/src/app_structure.rs [hunk 1]
```diff
@@ -1035,54 +1035,38 @@ async fn directory_tree_to_entrypoints_internal_untraced(
...
-                            ..Default::default()
-                        }.cell(),
+                        segment: "/_not-found".to_string(),
+                        parallel_routes: indexmap! {
+                            "children".to_string() => LoaderTree {
+                                page: app_page.clone(),
+                                segment: "__PAGE__".to_string(),
+                                parallel_routes: IndexMap::new(),
+                                components: Components {
+                                    page: if components.not_found.is_some() { components.not_found } else { Some(get_next_package(app_dir).join("dist/client/components/not-found-error.js".to_string())) },
```
[the remark this answers is not in the document]
Comment (timneutkens) [reply packages/next-swc/crates/next-core/src/app_structure.rs]: Editor already runs fmt and just manually ran `cargo fmt --package next-core` as well as `cargo fmt --all` and it doesn't update this line, not sure why.

**Review by huozhi — APPROVED** [review]

File: packages/next/src/export/routes/app-page.ts [hunk 2]
```diff
@@ -41,6 +41,7 @@ export async function exportAppPage(
   fileWriter: FileWriter
 ): Promise<ExportRouteResult> {
   // If the page is `/_not-found`, then we should update the page to be `/404`.
+  // UNDERSCORE_NOT_FOUND_ROUTE value used here, however we don't want to import it here as it causes constants to be inlined which we don't want here.
```
Comment (huozhi) [inline packages/next/src/export/routes/app-page.ts:44]: We could probably separate this into some other file to keep the bundle small later. not blocking

**Pull request merged**